File: problems/821-shortest-distance-to-a-character/binary-search.cpp
/**
 * 二分查找
 * 
 * 时间：`O(NlogM)`. 4ms
 */

class Solution {
public:
    vector<int> shortestToChar(string S, char pivot) {
        const int n = S.size();

        vector<int> indices; // 保存pivot的所有下标（升序）
        for (int i = 0; i < n; ++i) {
            if (S[i] == pivot) {
                indices.push_back(i);
            }
        }

        vector<int> res(n);
        for (int i = 0; i < n; ++i) {
            if (S[i] == pivot) {
                res[i] = 0;
            } else {
                res[i] = getMinDist(indices, i);
            }
        }
        return res;
    }

    // 二分查找：寻找离`target`最近的下标
    // 返回距离
    int getMinDist(vector<int> &indices, int target) {
        int left = 0, right = indices.size();
        while (left < right) {
            int mid = left + (right - left >> 1);
            if (target > indices[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        // 两个候选值，取距离更小者
        int res = INT_MAX;
        if (left < indices.size()) {
            res = min(res, abs(indices[left] - target));
        }
        if (left - 1 >= 0) {
            res = min(res, abs(indices[left - 1] - target));
        }
        return res;
    }
};

File: 剑指offer/11/-BinarySearch.js
/**
 * 二分查找
 */
var minArray = function (numbers) {
  let left = 0;
  let right = numbers.length;
  const pivot = numbers[0]; // 重点：选首元素作为二分的参照物

  while (left < right) {
    const mid = left + (right - left >>> 1);
    if (numbers[mid] === pivot) { // 相等 => 

    } else if (numbers[mid] < pivot) { // 比首元素小，
      right = mid;
    } else { // 大于等于首元素，往右走（mid不是答案）
      left = mid + 1;
    }
  }

  if (left >= numbers.length) { // 不是旋转数组的情况（left会一直往右走，直到超过数组长度）
    return numbers[0];
  }
  return numbers[left];
};